caffe2/core/dispatch/TensorTypeId.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include "caffe2/utils/IdWrapper.h"
+#include <string>
+#include <iostream>
+#include <mutex>
+#include <unordered_set>
+
+namespace c10 {
+
+namespace details {
+  using _tensorTypeId_underlyingType = uint8_t;
```

Comment:
it's a little fancy code to do registration. What C2 does is just have a single Enum of DeviceType (which is equivalent of TypeId here). The rationale is that you'd need to serialize them at some point too and there won't be too many values.

Dynamic registration is cool and I think it shouldn't cause issues, but it does create a lot of boilerplate for something that could be an enum for some time :)

Reply:
Just double checking, is it intended to capture just the device (aka, CPU, CUDA, ...), or device AND data type (aka, mapping the tuple (CPU, float) to a type id)?

If it is just device types, I agree with Dima that it should probably be an enum. The reason is that, we are going to be doing runtime registration with potentially an unset order, and as a result, since we need to do serialization, it is possible that people compare apples with oranges, for example comparing a DeviceType Enum with a dynamically allocated tensor type id.

For reading purposes, this is how Caffe2 defines devices with a protobuf enum and map them to names (for debugging purposes):

https://github.com/pytorch/pytorch/blob/e5b997223ccbc50373e3c53f6bfe58fe9d4efc06/caffe2/utils/proto_utils.cc#L55-L81
https://github.com/pytorch/pytorch/blob/e5b997223ccbc50373e3c53f6bfe58fe9d4efc06/caffe2/proto/caffe2.proto#L113-L124

this is how onnx defines device types with an enum and a map with names:
https://github.com/pytorch/pytorch/blob/a2a28c0ef1d9a433972fe72fa5b0b9b850ccfcaf/caffe2/onnx/device.cc#L7-L10

and this is how torch uses enum:
https://github.com/pytorch/pytorch/blob/749d51414af11da48bdb195375f702822450aded/torch/csrc/utils/device.h#L7

Note that the enum value assignment and the actual implementation can be separate: for example you can have a CUDA enum value defined but you do not have to have cuda built in.